packages/phishing-controller/src/PhishingDetector.ts

```diff
@@ -0,0 +1,177 @@
+import { distance } from 'fastest-levenshtein';
+
+import {
+  domainPartsToDomain,
+  domainPartsToFuzzyForm,
+  domainToParts,
+  getDefaultPhishingDetectorConfig,
+  matchPartsAgainstList,
+  processConfigs,
+} from './utils';
+
+export type LegacyPhishingDetectorList = {
+  whitelist?: string[];
+  blacklist?: string[];
+} & FuzzyTolerance;
+
+export type PhishingDetectorList = {
+  allowlist?: string[];
+  blocklist?: string[];
+  name?: string;
+  version?: string | number;
+} & FuzzyTolerance;
+
+export type FuzzyTolerance =
+  | {
+      tolerance?: number;
+      fuzzylist: string[];
+    }
+  | {
+      tolerance?: never;
+      fuzzylist?: never;
+    };
+
+export type PhishingDetectorOptions =
+  | LegacyPhishingDetectorList
+  | PhishingDetectorList[];
+
+export type PhishingDetectorConfiguration = {
+  name?: string;
+  version?: number | string;
+  allowlist: string[][];
+  blocklist: string[][];
+  fuzzylist: string[][];
+  tolerance: number;
+};
+
+export type PhishingDetectorResult = {
+  name?: string;
+  version?: string;
+  result: boolean;
+  match?: string;
+  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist' | 'blacklist' | 'whitelist';
```

Comment:
This is a union between the legacy response and the new one. 

There is probably a better way to do this, and it most likely involves writing two different types of responses (one for legacy, one for the new one) and overloading the `check` method - but as it depends on how the class was instantiated we might need a generic type at the class level to know which of the two types the `check` method will return

Reply:
I agree, the legacy config support makes the detector logic a bit overcomplicated since it just changes the way the list is returned, and a breaking change here to cut the support would make more sense. 

But since there are some changes over the original code due to the TS migration, I would keep this for a follow-up PR, where we can cut the legacy config support.

As for the server-side tooling to update the config (and the config itself), that will stay in the original repo

Reply:
TIL! So the answer is actually "yes", but fully agreed such refactors are best done separately.

> This makes me think: If the check method returns different results depending on how the PhishingDetector is instantiated, should we have two PhishingDetectors? Almost like a PhishingDetectorV1 and PhishingDetectorV2, or a LegacyPhishingDetector and PhishingDetector. We can do that in a different PR of course, but just some food for thought.

@mcmire So due to the code in this repo no longer being concerned with the serving side, sounds like the door is even open for fully removing what have been `PhishingDetectorV1`!

Reply:
Okay cool, that's good to know! I'm okay with the changes here and with removing V1 in another PR.